### DISCUSSION TITLE
Concurrency in Cheshire Cat AI: Is it designed for single-user or multi-user environments?

### DISCUSSION THREAD

**Discussion by rognoni** [cheshire-cat-ai/core#985] [Q&A]
👋 
I noticed that the agents in the **Cheshire Cat AI** project use LangChain's synchronous [invoke()](https://github.com/cheshire-cat-ai/core/blob/f14f2c0acebddd3f1f1efb6ba3b161bea45e9e3a/core/cat/agents/procedures_agent.py#L131) function instead of the asynchronous [ainvoke()](https://python.langchain.com/docs/concepts/async/#langchain-asynchronous-apis) function. Does this mean that the project is designed primarily for single-user, local usage, rather than for multiple concurrent users on a server?

I think it would be beneficial to discuss the design choices made in the project and explore if it's possible to adapt **Cheshire Cat AI** for handling multiple concurrent users efficiently.

**Comment by pieroit** [comment]
@rognoni every call to the agent and LLM is handled as a virtual thread (see fastAPI threadpool), no need for async.

In case you find issues about concurrency, let us know yo our knowledge there are none*

*Use Qdrant container instead of the sqlite version
